### PR TITLE
[198] Add password reset... and add magic link auth

### PR DESCRIPTION
**Pull request by travjenkins**
## Changes

Initially this wast to add password reset. As I dove into handling the edge cases it seems like just using magic link auth is MUCH easier. I will leave the password reset stuff in there on the random off chance someone gets to it... somehow.

Moving towards the `notistack` library for notifications. Right now just login is using it. But next we'll convert the notification store to fire off requests to notistack.

Added an image to the Magic Link email

## Tests

Manual

## Issues

Fixes: https://github.com/estuary/ui/issues/198

## Content

Some new messages around logging in

## Screenshots

Edge case... someone gets password reset link
image

Missing email
image

Supabase returns error trying to send email
image

OIDC for Google failed
image

Not allowed to user magic link multiple times. We'll log you out if you try
image

Limit on number of magic link requests
image

Magic Link Email
image

### CODE REVIEW

File: src/app/Authenticated.tsx [hunk 1]
```diff
@@ -76,9 +76,14 @@ export const routeDetails = {
...
-        title: 'routeTitle.registration',
-        path: '/register',
+    user: {
+        title: 'routeTitle.user',
+        path: '/user',
+        registration: {
+            title: 'routeTitle.registration',
+            path: 'register',
+            fullPath: '/user/register',
+        },
```
Comment (travjenkins) [inline src/app/Authenticated.tsx:86]: Initially this contained password reset but that was removed. I still think we'll need some kind of "user" path so leaving it

File: src/app/Authenticated.tsx [hunk 2]
```diff
@@ -98,15 +103,21 @@ const Authenticated = () => {
...
+            <Routes>
+                <Route path={routeDetails.user.path}>
                     <Route
-                        path={routeDetails.registration.path}
+                        path={routeDetails.user.registration.path}
                         element={<Registration />}
                     />
+                </Route>
+                <Route path="*" element={<NoGrantsFound />} />
+            </Routes>
```
Comment (travjenkins) [inline src/app/Authenticated.tsx:115]: Making folks that hit "No Grants" able to land on the registration page as that will be who we need to "register"

File: src/app/Unauthenticated.tsx [hunk 3]
```diff
@@ -2,19 +2,12 @@ import Login from 'pages/Login';
 import { Route, Routes } from 'react-router';
 
 export const logoutRoutes = {
-    path: 'logout',
-    params: {
-        reason: 'reason',
-    },
+    path: '',
```
Comment (travjenkins) [inline src/app/Unauthenticated.tsx:5]: Redid the reason thing cause it was a major pain and kinda weird to make the login page a notification controller.

File: src/components/login/MagicLink.tsx [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import React, { useState } from 'react';
+import { FormattedMessage, useIntl } from 'react-intl';
+import {
+    defaultOptions,
+    defaultRenderers,
+    hideValidation,
+    showValidation,
+} from 'services/jsonforms';
+
+const MagicLink = () => {
```
Comment (travjenkins) [inline src/components/login/MagicLink.tsx:17]: New form to handle submitting magic link requests.

File: src/components/login/MagicLink.tsx [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+                description: intl.formatMessage({
+                    id: 'login.email.description',
+                }),
+                minLength: 5,
+                type: 'string',
+            },
+        },
+        required: ['email'],
+        type: 'object',
+    };
```
Comment (travjenkins) [inline src/components/login/MagicLink.tsx:43]: We might want to eventually store away common "specs" that could end up being used elsewhere.

File: src/components/login/MagicLink.tsx [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        );
+    };
+
+    const handlers = {
+        submit: async (event: React.FormEvent<HTMLFormElement>) => {
+            event.preventDefault();
+            if (!isEmpty(formErrors) || isEmpty(formData.email)) {
+                setShowErrors(true);
+                return;
+            }
```
Comment (travjenkins) [inline src/components/login/MagicLink.tsx:79]: Starting to like the pattern when using async of checking for errors and then returning if there are some.

File: src/components/login/MagicLink.tsx [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            setSubmitError(null);
+            setShowErrors(false);
+            setLoading(true);
+
+            const { error } = await supabaseClient.auth
+                .signIn({
+                    email: formData.email,
+                })
+                .finally(() => {
+                    setLoading(false);
```
Comment (travjenkins) [inline src/components/login/MagicLink.tsx:90]: Need to make sure we set off loading no matter what.

File: src/components/login/OIDCs.tsx [hunk 8]
```diff
@@ -1,27 +1,61 @@
...
     const redirectTo = useConstant(
         () => `${window.location.origin}` // `${window.location.origin}${routeDetails.registration.path}`
     );
 
+    const supabaseClient = useClient();
+    const intl = useIntl();
+
+    const { enqueueSnackbar } = useSnackbar();
+
+    const loginFailed = (key: Provider) => {
```
Comment (travjenkins) [inline src/components/login/OIDCs.tsx:19]: All OIDCs can use the same general pattern when messaging that they were unable to login.

File: src/components/login/OIDCs.tsx [hunk 9]
```diff
@@ -1,27 +1,61 @@
...
             sx={{
                 display: 'flex',
                 justifyContent: 'center',
             }}
         >
-            <GoogleAuthButton
-                onError={() => onError('login.loginFailed.google')}
-                redirectPath={redirectTo}
-            />
+            <GoogleButton onClick={() => login('google')} />
```
Comment (travjenkins) [inline src/components/login/OIDCs.tsx:58]: The wrapper that was there no longer was serving any purpose.

File: package.json [hunk 10]
```diff
@@ -42,6 +42,7 @@
         "logrocket-react": "^5.0.1",
         "lru-cache": "^7.10.1",
         "monaco-editor": "^0.33.0",
+        "notistack": "^2.0.5",
```
Comment (travjenkins) [inline package.json:45]: New library for managing displaying notifications.

File: src/context/SWR.tsx [hunk 11]
```diff
@@ -18,27 +16,32 @@ export const DEFAULT_POLLING = 2500;
...
+                            id: 'login.jwtExpired',
+                        }),
+                        {
+                            anchorOrigin: {
+                                vertical: 'top',
+                                horizontal: 'center',
+                            },
+                            variant: 'error',
+                        }
                     );
```
Comment (travjenkins) [inline src/context/SWR.tsx:45]: Just making the thing that is forcing the log out responsible for the notification. Makes a lot of stuff simpler. Also, with notistack we can have references to what opened it and put in custom actions so this empowers some interesting possibilities.

File: src/context/User.tsx [hunk 12]
```diff
@@ -1,9 +1,39 @@
...
+                        id: 'login.passwordReset',
+                    }),
+                    {
+                        anchorOrigin: {
+                            vertical: 'top',
+                            horizontal: 'center',
+                        },
+                        variant: 'info',
+                    }
+                );
```
Comment (travjenkins) [inline src/context/User.tsx:30]: Super weird edge case but might as well catch it. If a user got a recovery email link just adding a notification that they should not need to do that.

File: src/context/User.tsx [hunk 13]
```diff
@@ -1,9 +1,39 @@
...
+            } else if (event === 'SIGNED_OUT') {
+                navigate(logoutRoutes.path);
```
Comment (travjenkins) [inline src/context/User.tsx:32]: If the user signed out in any way navigate to the root path.

File: src/context/User.tsx [hunk 14]
```diff
@@ -1,9 +1,39 @@
...
+                    {
+                        anchorOrigin: {
+                            vertical: 'top',
+                            horizontal: 'center',
+                        },
+                        variant: 'info',
+                    }
+                );
+            } else if (event === 'SIGNED_OUT') {
+                navigate(logoutRoutes.path, { replace: true });
```
Comment (travjenkins) [inline src/context/User.tsx:32]: If someone is logged out somehow we want to force them to the login path and replace any path they might be at.

File: src/services/ajv.ts [hunk 15]
```diff
@@ -1,22 +1,25 @@
 import { createAjv } from '@jsonforms/core';
 
+// TODO (typing) Need to get this typed as the AJV Options type
+export const defaultAjvSettings: any = {
```
Comment (travjenkins) [inline src/services/ajv.ts:4]: Need the default settings else where so making available

File: src/services/supabase.ts [hunk 16]
```diff
@@ -21,6 +21,10 @@ const supabaseSettings = {
     anonKey: process.env.REACT_APP_SUPABASE_ANON_KEY,
 };
 
+// Related to the AUth stuff
+export const supabaseAuthError = 'error_description';
+export const userNotFound = 'User not found';
```
Comment (travjenkins) [inline src/services/supabase.ts:26]: I want to find a better way to know these... might look into using the library from Supabase directly? Not sure yet.

**Review by kiahna-tucker — APPROVED** [review]
